### PR TITLE
feat: Support multi-modal input and multi-modal output in one agent

### PR DESCRIPTION
**Pull request by Zhoues**
## Description

Mainly implements 4 multi-modal parts:

1. Building an agent to support multiple image inputs (see examples/vision/object_recognition.py)
2. Building an agent to call DALL-E to generate images (see examples/vision/image_crafting.py)
3. Building an agent to support calls to DALL-E to generate images while understanding multiple image inputs (see examples/vision/multi_condition_image_crafting.py)
4. Build two agents one for generating the image and one for suggesting ways for the other to fix the image. (see examples/vision/multi_turn_image_refining.py)

## Motivation and Context

Part of #454 

also will fix https://github.com/camel-ai/camel/issues/541

https://github.com/camel-ai/camel/issues/324

- [x] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [x] Example (update in the folder of example)

## Implemented Tasks

- Supports multi-image input
- Support DALLE function call
- To allow an agent to receive image inputs and output images at the same time, a new `FunctionCallingVisionConfig` is built to adapt the `ChatGPTVisionConfig`
- Create image crafting task type and example
- Create multi-condition  image crafting task type and example

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by yiyiyi0817 — COMMENTED** [review]
It appears that your code hasn't passed the CI/CD tests. I kindly suggest taking a look at https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md for guidance. Once you've resolved the formatting issues, mypy checks, and ensured that pytest is successful, hope you could resubmit your code. Thank you! @Zhoues

File: camel/prompts/image_craft.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+from typing import Any
+
+from camel.prompts import TextPrompt, TextPromptDict
+from camel.types import RoleType
+
+
+class ImageCraftPromptTemplateDict(TextPromptDict):
+    ASSISTANT_PROMPT = TextPrompt(
+        """You are tasked with creating an original image based on the provided descriptive captions. Please use your imagination and artistic capabilities to visualize and draw the images and explain what you are thinking about.""")
```
Comment (zechengz) [inline camel/prompts/image_craft.py:22]: nit:
```suggestion
        """You are given the task of generating an original image based on the descriptive captions. Please use your creativity and artistic skills to visualize and create an image with your thought process.""")
```

**Comment by Wendong-Fan** [comment]
Hey @Zhoues , whether it's ok for you to let other contributor working on this PR?

**Comment by Zhoues** [comment]
> Hey @Zhoues , whether it's ok for you to let other contributor working on this PR?

Of course, I'd be more than happy to let other contributors work on this PR！

**Comment by willshang76** [comment]
Close this, since there is a new PR for people to keep working on this.

**Pull request closed without merging**